src/main/java/io/github/thebusybiscuit/slimefun4/api/researches/Research.java

```diff
@@ -209,24 +209,29 @@ public List<SlimefunItem> getAffectedItems() {
      *            The {@link PlayerProfile} of that {@link Player}.
      * @param sfItem
      *            The {@link SlimefunItem} on which the {@link Player} clicked.
-     * @param itemGroup
-     *            The {@link ItemGroup} where the {@link Player} was.
-     * @param page
-     *            The page number of where the {@link Player} was in the {@link ItemGroup};
+     * @param callback
+     *            The action to perform after unlocking this item.
      *
      */
     @ParametersAreNonnullByDefault
-    public void unlockFromGuide(SlimefunGuideImplementation guide, Player player, PlayerProfile profile, SlimefunItem sfItem, ItemGroup itemGroup, int page) {
+    public void unlockFromGuide(SlimefunGuideImplementation guide, Player player, PlayerProfile profile, SlimefunItem sfItem, Consumer<Player> callback) {
         if (!Slimefun.getRegistry().getCurrentlyResearchingPlayers().contains(player.getUniqueId())) {
             if (profile.hasUnlocked(this)) {
-                guide.openItemGroup(profile, itemGroup, page);
+                callback.accept(player);
             } else {
                 PlayerPreResearchEvent event = new PlayerPreResearchEvent(player, this, sfItem);
                 Bukkit.getPluginManager().callEvent(event);
 
                 if (!event.isCancelled()) {
                     if (this.canUnlock(player)) {
-                        guide.unlockItem(player, sfItem, pl -> guide.openItemGroup(profile, itemGroup, page));
+                        if (player.getGameMode() == GameMode.CREATIVE && Slimefun.getRegistry().isFreeCreativeResearchingEnabled()) {
+                            unlock(player, true, callback);
+                        } else {
+                            player.setLevel(player.getLevel() - this.getCost());
+
+                            boolean skipLearningAnimation = Slimefun.getRegistry().isLearningAnimationDisabled() || !SlimefunGuideSettings.hasLearningAnimationEnabled(player);
+                            unlock(player, skipLearningAnimation, callback);
+                        }
```

Comment:
if the player is in creative, but freeCreativeResearchingnabled is disabled, it wont try to unlock with their actual levels
is that intended?

Reply:
It will. Only if the player is in creative mode and free researching is enabled, the research will be unlocked immediately. Otherwise it will cost xp levels.

Besides, that code was only moved here from a method which we didnt really need. The logic itself remained unchanged.

Reply:
yeah I saw it was copied but I was wondering it anyway, but you're right it will get fired, idk why I thought it wouldnt